_posts/2021-01-22-x-ways-to-write-meaningful-commit-messages.md

```diff
@@ -0,0 +1,140 @@
+
+If you fix a bug and write: "Fixed the bug!". What happens is that you will probably completely forget about it a month or two later. So it's not feasible to think if it's not clear for others, they can ask about it. It can waste a lot of time and energy as well.
+
+## What is a good commit message?
+
+Writing suitable commit messages can be defined as many styles. The trick is to pick the best that suits you and then stick to it. It means being **consistent** in using the format in all commits.
+
+The perfect commit message should have certain qualities:
+  - **Understandable** even by seeing only the header of the message (We'll talk about the header soon)
+  - Not too much detailed, but **just enough**
+  - **Unambiguous**
+
+Writing suitable commit messages can be defined as many models. The trick is to pick the best that suits you and then stick to it. It means being consistent in using the format in all commits.
+
+There are many styles we can follow to reach these standards. Angular suggested style is one of them that we're going to explain a simplified version of it later in this article.
+
+All styles and our commits should have some specific qualities to form a structured commit message which we’re going to explain now.
+
+### Atomic Commit:
+Although using a proper style is a good practice, but it's not enough. The discipline is crucial. The commits should be **reasonable and atomic**. If the commit consists of multiple changes that make the message too long or inefficient, it's good practice to **separate it into several commits**. In other words, you don't want to commit a code that does nothing or does too much, or maybe it's broken.
+
+If you commit two changes together, for example, a bug fix, and a minor refactoring, It might not cause a very long commit message, but it can cause some problems. Let's say the bug fix made some other bugs. In that case, you should roll back to your recent commit. It results in the loss of that refactoring as well. It's not efficient, and It's not atomic.
+
+### Short and Unambiguous:
+The message should describe what changes your code makes to the base code. It should answer the question: **"What happens if the changes are applied?"**. If you can not be short, it might be because the commit is not atomic, and it's too much change in one commit.
+
+#### Active Voice:
+ Use the **imperative, present tense**. It is easier to read and skim.
+> **Right**: Add feature for a user to upload pictures
+>
+> **Wrong**: Added feature ... (passive voice, past tense)
+
+#### Detailed Enough:
+Super-detailed commit messages are frustrated as well. You can find that much detail in the system. For example, If your version control is Git, you can see all the changed files in Git. So instead of answering "what are the changes? ", it's better to answer **"What are the changes for?"**.
+
+### Formatting:
```

Comment:
It would be nice if we compared multiple different formatting styles. Otherwise, the title of this article would be lying :).

For example:
* what are semantic commit messages? 
* what are "conventional commits" (https://www.conventionalcommits.org/en/v1.0.0/)? I think this is an evolution of semantic commit messages
* then, describe the Angular style as an instance of conventional commits
* are there other styles? if yes, describe them shortly.
* do some more research and maybe link to helpful articles like this one: https://medium.com/@auscunningham/enforcing-git-commit-message-style-b86a45380b0f

Reply:
Sure. From the articles, I found out there can be multiple ways to write a commit message, and as long as all have some qualities, they're all great. So I though "X" ways kind of refers to this fact. But all you've mentioned where the questions I have struggled with them as well. So thanks for the links. I read them.
One question though:
If for example I mention 3 different ways, should I change the title to "3 ways to write commit messages" as well ?

Reply:
We might change the title completely in the end and strip out the "X ways". Don't worry about that, I'll change the title in the end.